src/engine/Join.cpp

```diff
@@ -762,77 +763,56 @@ void Join::hashJoin(const IdTable& dynA, size_t jc1, const IdTable& dynB,
 
   IdTableStatic<OUT_WIDTH> result = dynRes->moveToStatic<OUT_WIDTH>();
 
-  // Cannot just switch l1 and l2 around because the order of
+  // Puts the rows of the given table into a hash map, with the value of
+  // the join column of a row as the key, and returns the hash map.
+  auto putIntoHashMap = [] (const auto table, const size_t jc, const auto WIDTH) {
+    // This declaration works, because generic lambdas are just syntactic sugar
+    // for templates. So, if the parameter value of WIDTH is avaible at
+    // compilet ime (and marked as such with std::integral_constant), we can
+    // use WIDTH for a template parameter.
+    ad_utility::HashMap<Id, std::vector<std::decay_t<typename IdTableStatic<WIDTH>::const_row_type>>> map;
+    for ( size_t j = 0; j < table.size(); j++) {
+      map[table(j, jc)].push_back(table[j]);
+    }
+    return map;
+  };
+
+  // Cannot just switch a and b around because the order of
   // items in the result tuples is important.
-  if (a.size() / b.size() > GALLOP_THRESHOLD) {
-    doGallopInnerJoin(LeftLargerTag{}, a, jc1, b, jc2, &result);
-  } else if (b.size() / a.size() > GALLOP_THRESHOLD) {
-    doGallopInnerJoin(RightLargerTag{}, a, jc1, b, jc2, &result);
   // Procceding with the actual hash join depended on which IdTableView
   // is bigger.
-  } else if (a.size() / b.size() >= 1) {
+  if (a.size() >= b.size()) {
     // a is bigger, or the same size as b. b gets put into the hash table.
-    ad_utility::HashMap<Id, std::vector<std::decay_t<typename IdTableStatic<R_WIDTH>::const_row_type>>> map;
-    for ( size_t j = 0; j < b.size(); j++) {
-      map[b(j, jc2)].push_back(b[j]);
-    }
+    auto map = putIntoHashMap(b, jc2, std::integral_constant<int, R_WIDTH>{});
+
 
     // Create cross product by going through a.
     for (size_t i = 0; i < a.size(); i++) {
       // Skip, if there is no matching entry in the join column.
-      if (map.find(a(i, jc1)) == map.end()) {
+      auto entry = map.find(a(i, jc1));
```

Comment:
If the range-based for loop works above, then you can also apply it here (and for the similar case for `b`).

Reply:
Am I misunderstanding, or didn't I already implement this?

Reply:
Seems to be already implemented.